vertx-mysql-client/src/test/java/io/vertx/mysqlclient/MySQLPipeliningTest.java

```diff
@@ -0,0 +1,304 @@
+      .executeBatch(batchParams)
+      .onComplete(ctx.asyncAssertSuccess(res -> {
+        for (int i = 0; i < 1000; i++) {
+          ctx.assertEquals(1, res.rowCount());
+          res = res.next();
+          latch.countDown();
+        }
+
+        client.query("SELECT id, val FROM mutable")
+          .execute()
+          .onComplete(ctx.asyncAssertSuccess(res2 -> {
+            ctx.assertEquals(1000, res2.size());
+            int i = 0;
+            for (Row row : res2) {
+              ctx.assertEquals(2, row.size());
+              ctx.assertEquals(i, row.getInteger(0));
+              ctx.assertEquals(String.format("val-%d", i), row.getString(1));
+              i++;
+            }
+            client.close();
+          }));
+      }));
+  }
+
+  @Test
+  public void testBatchInsertExceptionConn(TestContext ctx) {
+    MySQLConnection.connect(vertx, options)
+      .onComplete(ctx.asyncAssertSuccess(conn -> {
+        testBatchInsertException(ctx, conn);
+      }));
+  }
+
+  @Test
+  public void testBatchInsertExceptionPool(TestContext ctx) {
+    SqlClient client = MySQLPool.client(vertx, options, new PoolOptions().setMaxSize(8));
+    testBatchInsertException(ctx, client);
+  }
+
+  private void testBatchInsertException(TestContext ctx, SqlClient client) {
+    List<Tuple> batchParams = new ArrayList<>();
+    for (int i = 0; i < 1000; i++) {
+      batchParams.add(Tuple.of(i, String.format("val-%d", i)));
+    }
+    batchParams.add(501, Tuple.of(500, "error")); // primary key violation error occurs in the 501st iteration
+
+    client.preparedQuery("INSERT INTO mutable(id, val) VALUES (?, ?)")
+      .executeBatch(batchParams)
+      .onComplete(ctx.asyncAssertFailure(error -> {
+        ctx.assertEquals(MySQLBatchException.class, error.getClass());
+        MySQLBatchException mySQLBatchException = (MySQLBatchException) error;
+        ctx.assertTrue(mySQLBatchException.getIterationError().containsKey(501));
+
+        // all the param will be executed
+        client.query("SELECT id, val FROM mutable")
+          .execute()
+          .onComplete(ctx.asyncAssertSuccess(res2 -> {
+            ctx.assertEquals(1000, res2.size());
+            int i = 0;
+            for (Row row : res2) {
+              ctx.assertEquals(2, row.size());
+              ctx.assertEquals(i, row.getInteger(0));
+              ctx.assertEquals(String.format("val-%d", i), row.getString(1));
+              i++;
+            }
+            client.close();
+          }));
+      }));
+  }
+
+  private void cleanTestTable(TestContext ctx) {
+    MySQLConnection.connect(vertx, options, ctx.asyncAssertSuccess(conn -> {
+      conn.query("TRUNCATE TABLE mutable;").execute(ctx.asyncAssertSuccess(result -> {
+        conn.close();
+      }));
+    }));
+  }
+
+  private void testSequentialQuery(TestContext ctx, Function<Integer, Future<RowSet<Row>>> resultExecution) {
+    Async latch = ctx.async(1000);
+    for (int i = 0; i < 1000; i++) {
```

Comment:
I reworked the tests and it seems we have some issues with the ordering of pool connection aquirements

Reply:
do you mean the failure observed in MySQL tests are general and not related to MySQL?

Reply:
yes it should be a general pool issue, I have moved the pipelining query tests to TCK in https://github.com/eclipse-vertx/vertx-sql-client/pull/1168/commits/3204c0cc19f676dbb00eb221e52a75115eb505e1, it does reproduce the same behaviors in both clients.

Reply:
I'll have a look

Reply:
this is actually the correct behaviour I'm observing, the SQL client has a 8 pooled connections, so the queries will be load balanced on the 8 connections and the result will be un-ordered.

the test could pass with a single connection instead of 8

do you think this test makes sense ?

Reply:
you are right, the pipelined queries should be executed on the pool out of order, I heve fixed the incorrect testcases